[Automation] java run fails for org.specs2:specs2-fp_3:5.0.0

## java run on ubuntu-latest / JDK 25 / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: 25
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="5.0.0" ./gradlew clean javaTest -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25628675165/job/75228360837
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 9s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 5.0.0 at 2026-05-10T12:39:17Z with exit code 1 after 9s
❌ ERROR: 'native-image run' for 5.0.0 failed with exit code 1.
SOFT-FAIL[native-image run][5.0.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="5.0.0" ./gradlew clean test -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"]
========================================================================================
 org.specs2:specs2-fp_3:4.20.5:5.0.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 5.0.0 at 2026-05-10T12:39:17Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="5.0.0" ./gradlew clean compileTestJava -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 745ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 776ms
9 actionable tasks: 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 5.0.0 at 2026-05-10T12:39:20Z with exit code 0 after 3s
========================================================================================
 org.specs2:specs2-fp_3:4.20.5:5.0.0 stage java run
========================================================================================
Starting stage 'java run' for 5.0.0 at 2026-05-10T12:39:20Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="5.0.0" ./gradlew clean javaTest -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 743ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :compileScala NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :tck-test-support:jar UP-TO-DATE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

> Task :compileTestScala
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:14:22: value EitherIdOps is not a member of org.specs2.fp
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:15:22: value EitherObjectOps is not a member of org.specs2.fp
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:16:22: value EitherOps is not a member of org.specs2.fp
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:45:43: [33mobject[0m [35mMonad[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:49:77: Could not infer type for parameter i of anonymous function

Expected type for the whole anonymous function: Int
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:68:46: value whenM is not a member of org.specs2.fp.Applicative[Option] - did you mean optionApplicative.when?
exception caught when loading module class ModuleUtils$: java.lang.AssertionError: org.junit.platform.commons.util.ModuleUtils$ModuleReferenceScanner
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:69:46: value unlessM is not a member of org.specs2.fp.Applicative[Option] - did you mean optionApplicative.unless?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:159:53: No given instance of type org.specs2.fp.Monad[[_] =>> Any] was found for a context parameter of [33mmethod[0m [35mfoldLeftM[0m in [33mtrait[0m [35mFoldableSyntax[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:160:57: No given instance of type org.specs2.fp.Applicative[[_] =>> Any] was found for a context parameter of [33mmethod[0m [35mtraverse_[0m in [33mtrait[0m [35mFoldableSyntax[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:168:48: [33mobject[0m [35mTraverse[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:169:52: [33mobject[0m [35mTraverse[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:170:71: [33mobject[0m [35mTraverse[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:177:67: No given instance of type org.specs2.fp.Applicative[[_] =>> Any] was found for a context parameter of [33mmethod[0m [35mtraverse[0m in [33mtrait[0m [35mTraverseSyntax[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:191:23: Not found: type EitherObjectOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:191:45: Not found: type EitherObjectOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:192:19: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:192:48: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:193:18: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:193:47: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:226:32: Not found: type EitherIdOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:227:39: Not found: type EitherIdOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:287:33: Found:    [1m[31mStream[0m[org.specs2.fp.Tree[String]]
Required: [1m[32mLazyList[0m[org.specs2.fp.Tree[String]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:296:29: No given instance of type org.specs2.fp.Show[String] was found for parameter sh of [33mmethod[0m [35mdrawTree[0m in [33mclass[0m [35mTree[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:304:99: Found:    [1m[31mStream[0m[org.specs2.fp.Tree[String]]
Required: [1m[32mLazyList[0m[org.specs2.fp.Tree[String]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:323:40: value fromForest is not a member of object org.specs2.fp.TreeLoc
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:332:33: Found:    [1m[31mStream[0m[org.specs2.fp.Tree[String]]
Required: [1m[32mLazyList[0m[org.specs2.fp.Tree[String]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:339:64: Could not infer type for parameter x$1 of anonymous function

In expanded function:
x$1 =>
  x$1[33m match [0m
    {
      [33mcase [0m(label, length) =>
        (label, length)
    }

Expected type for the whole anonymous function: Int
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:339:89: Not found: label - did you mean labelled?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:339:96: Not found: length
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:350:20: Found:    (children : Stream[Int])
Required: LazyList[Int]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:352:69: Found:    Stream[Int]
Required: LazyList[Any]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:355:49: Found:    LazyList[Int]
Required: Stream[Int]
32 errors found

> Task :compileTestScala FAILED


FAILURE: Build failed with an exception.

* What went wrong:
[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
12 actionable tasks: 2 executed, 10 up-to-date
Execution failed for task ':compileTestScala'.
> Compilation failed

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 6s

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 8s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 5.0.0 at 2026-05-10T12:39:28Z with exit code 1 after 8s
❌ ERROR: 'java run' for 5.0.0 failed with exit code 1.
FAILED[java run][5.0.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="5.0.0" ./gradlew clean javaTest -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"]
```

## java run on ubuntu-latest / JDK latest-ea / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="5.0.0" ./gradlew clean javaTest -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25628675165/job/75228360836
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
adle documentation.

BUILD FAILED in 9s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 5.0.0 at 2026-05-10T12:39:58Z with exit code 1 after 9s
❌ ERROR: 'native-image run' for 5.0.0 failed with exit code 1.
SOFT-FAIL[native-image run][5.0.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="5.0.0" ./gradlew clean test -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"]
========================================================================================
 org.specs2:specs2-fp_3:4.20.5:5.0.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 5.0.0 at 2026-05-10T12:39:58Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="5.0.0" ./gradlew clean compileTestJava -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 757ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 786ms
9 actionable tasks: 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 5.0.0 at 2026-05-10T12:40:00Z with exit code 0 after 2s
========================================================================================
 org.specs2:specs2-fp_3:4.20.5:5.0.0 stage java run
========================================================================================
Starting stage 'java run' for 5.0.0 at 2026-05-10T12:40:00Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="5.0.0" ./gradlew clean javaTest -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 761ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :compileScala NO-SOURCE
> Task :tck-test-support:jar UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

> Task :compileTestScala
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:14:22: value EitherIdOps is not a member of org.specs2.fp
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:15:22: value EitherObjectOps is not a member of org.specs2.fp
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:16:22: value EitherOps is not a member of org.specs2.fp
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:45:43: [33mobject[0m [35mMonad[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:49:77: Could not infer type for parameter i of anonymous function

Expected type for the whole anonymous function: Int
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:68:46: value whenM is not a member of org.specs2.fp.Applicative[Option] - did you mean optionApplicative.when?
exception caught when loading module class ModuleUtils$: java.lang.AssertionError: org.junit.platform.commons.util.ModuleUtils$ModuleReferenceScanner[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:69:46: value unlessM is not a member of org.specs2.fp.Applicative[Option] - did you mean optionApplicative.unless?

[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:159:53: No given instance of type org.specs2.fp.Monad[[_] =>> Any] was found for a context parameter of [33mmethod[0m [35mfoldLeftM[0m in [33mtrait[0m [35mFoldableSyntax[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:160:57: No given instance of type org.specs2.fp.Applicative[[_] =>> Any] was found for a context parameter of [33mmethod[0m [35mtraverse_[0m in [33mtrait[0m [35mFoldableSyntax[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:168:48: [33mobject[0m [35mTraverse[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:169:52: [33mobject[0m [35mTraverse[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:170:71: [33mobject[0m [35mTraverse[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:177:67: No given instance of type org.specs2.fp.Applicative[[_] =>> Any] was found for a context parameter of [33mmethod[0m [35mtraverse[0m in [33mtrait[0m [35mTraverseSyntax[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:191:23: Not found: type EitherObjectOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:191:45: Not found: type EitherObjectOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:192:19: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:192:48: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:193:18: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:193:47: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:226:32: Not found: type EitherIdOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:227:39: Not found: type EitherIdOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:287:33: Found:    [1m[31mStream[0m[org.specs2.fp.Tree[String]]
Required: [1m[32mLazyList[0m[org.specs2.fp.Tree[String]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:296:29: No given instance of type org.specs2.fp.Show[String] was found for parameter sh of [33mmethod[0m [35mdrawTree[0m in [33mclass[0m [35mTree[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:304:99: Found:    [1m[31mStream[0m[org.specs2.fp.Tree[String]]
Required: [1m[32mLazyList[0m[org.specs2.fp.Tree[String]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:323:40: value fromForest is not a member of object org.specs2.fp.TreeLoc
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:332:33: Found:    [1m[31mStream[0m[org.specs2.fp.Tree[String]]
Required: [1m[32mLazyList[0m[org.specs2.fp.Tree[String]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:339:64: Could not infer type for parameter x$1 of anonymous function

In expanded function:
x$1 =>
  x$1[33m match [0m
    {
      [33mcase [0m(label, length) =>
        (label, length)
    }

Expected type for the whole anonymous function: Int
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:339:89: Not found: label - did you mean labelled?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:339:96: Not found: length
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:350:20: Found:    (children : Stream[Int])
Required: LazyList[Int]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:352:69: Found:    Stream[Int]
Required: LazyList[Any]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:355:49: Found:    LazyList[Int]
Required: Stream[Int]
32 errors found

> Task :compileTestScala FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileTestScala'.
> Compilation failed

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 7s

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
12 actionable tasks: 2 executed, 10 up-to-date

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 9s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 5.0.0 at 2026-05-10T12:40:10Z with exit code 1 after 10s
❌ ERROR: 'java run' for 5.0.0 failed with exit code 1.
FAILED[java run][5.0.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="5.0.0" ./gradlew clean javaTest -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"]
```

## java run on ubuntu-latest / JDK latest-ea / mode future-defaults-all

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: future-defaults-all
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="5.0.0" ./gradlew clean javaTest -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25628675165/job/75228360842
Last 300 lines of the log (excerpt truncated to fit the GitHub issue size limit):
```console
n scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 9s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 5.0.0 at 2026-05-10T12:40:43Z with exit code 1 after 9s
❌ ERROR: 'native-image run' for 5.0.0 failed with exit code 1.
SOFT-FAIL[native-image run][5.0.0][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="5.0.0" ./gradlew clean test -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"]
========================================================================================
 org.specs2:specs2-fp_3:4.20.5:5.0.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 5.0.0 at 2026-05-10T12:40:43Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="5.0.0" ./gradlew clean compileTestJava -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 761ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 783ms
9 actionable tasks: 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 5.0.0 at 2026-05-10T12:40:46Z with exit code 0 after 3s
========================================================================================
 org.specs2:specs2-fp_3:4.20.5:5.0.0 stage java run
========================================================================================
Starting stage 'java run' for 5.0.0 at 2026-05-10T12:40:46Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="5.0.0" ./gradlew clean javaTest -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 766ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileScala NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava NO-SOURCE
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:14:22: value EitherIdOps is not a member of org.specs2.fp
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:15:22: value EitherObjectOps is not a member of org.specs2.fp
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:16:22: value EitherOps is not a member of org.specs2.fp

> Task :compileTestScala
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:45:43: [33mobject[0m [35mMonad[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:49:77: Could not infer type for parameter i of anonymous function

Expected type for the whole anonymous function: Int
exception caught when loading module class ModuleUtils$: java.lang.AssertionError: org.junit.platform.commons.util.ModuleUtils$ModuleReferenceScanner
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:68:46: value whenM is not a member of org.specs2.fp.Applicative[Option] - did you mean optionApplicative.when?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:69:46: value unlessM is not a member of org.specs2.fp.Applicative[Option] - did you mean optionApplicative.unless?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:159:53: No given instance of type org.specs2.fp.Monad[[_] =>> Any] was found for a context parameter of [33mmethod[0m [35mfoldLeftM[0m in [33mtrait[0m [35mFoldableSyntax[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:160:57: No given instance of type org.specs2.fp.Applicative[[_] =>> Any] was found for a context parameter of [33mmethod[0m [35mtraverse_[0m in [33mtrait[0m [35mFoldableSyntax[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:168:48: [33mobject[0m [35mTraverse[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:169:52: [33mobject[0m [35mTraverse[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:170:71: [33mobject[0m [35mTraverse[0m in [33mpackage[0m [35morg.specs2.fp[0m does not take type parameters
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:177:67: No given instance of type org.specs2.fp.Applicative[[_] =>> Any] was found for a context parameter of [33mmethod[0m [35mtraverse[0m in [33mtrait[0m [35mTraverseSyntax[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:191:23: Not found: type EitherObjectOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:191:45: Not found: type EitherObjectOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:192:19: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:192:48: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:193:18: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:193:47: Not found: type EitherOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:226:32: Not found: type EitherIdOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:227:39: Not found: type EitherIdOps
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:287:33: Found:    [1m[31mStream[0m[org.specs2.fp.Tree[String]]
Required: [1m[32mLazyList[0m[org.specs2.fp.Tree[String]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:296:29: No given instance of type org.specs2.fp.Show[String] was found for parameter sh of [33mmethod[0m [35mdrawTree[0m in [33mclass[0m [35mTree[0m
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:304:99: Found:    [1m[31mStream[0m[org.specs2.fp.Tree[String]]
Required: [1m[32mLazyList[0m[org.specs2.fp.Tree[String]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:323:40: value fromForest is not a member of object org.specs2.fp.TreeLoc
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:332:33: Found:    [1m[31mStream[0m[org.specs2.fp.Tree[String]]
Required: [1m[32mLazyList[0m[org.specs2.fp.Tree[String]]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:339:64: Could not infer type for parameter x$1 of anonymous function

In expanded function:
x$1 =>
  x$1[33m match [0m
    {
      [33mcase [0m(label, length) =>
        (label, length)
    }

Expected type for the whole anonymous function: Int
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:339:89: Not found: label - did you mean labelled?
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:339:96: Not found: length
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:350:20: Found:    (children : Stream[Int])
Required: LazyList[Int]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:352:69: Found:    Stream[Int]
Required: LazyList[Any]
[Error] /home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala:355:49: Found:    LazyList[Int]
Required: Stream[Int]
32 errors found

> Task :compileTestScala FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.specs2/specs2-fp_3/4.20.5/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileTestScala'.
> Compilation failed

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 7s
12 actionable tasks: 2 executed, 10 up-to-date

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 8s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 5.0.0 at 2026-05-10T12:40:55Z with exit code 1 after 9s
❌ ERROR: 'java run' for 5.0.0 failed with exit code 1.
FAILED[java run][5.0.0][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="5.0.0" ./gradlew clean javaTest -Pcoordinates="org.specs2:specs2-fp_3:4.20.5"]
```

--- a/metadata/org.specs2/specs2-fp_3/5.0.0/reachability-metadata.json
+++ b/metadata/org.specs2/specs2-fp_3/5.0.0/reachability-metadata.json
@@ -1,0 +1,43 @@
+{
+  "reflection": [
+    {
+      "condition": {
+        "typeReached": "org.specs2.fp.Need$$anon$2"
+      },
+      "type": "java.lang.StackTraceElement[]"
+    },
+    {
+      "condition": {
+        "typeReached": "org.specs2.fp.Monoid"
+      },
+      "type": "org.specs2.fp.Monoid$",
+      "fields": [
+        {
+          "name": "0bitmap$1"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "org.specs2.fp.Need$$anon$2"
+      },
+      "type": "org.specs2.fp.Need$$anon$2",
+      "fields": [
+        {
+          "name": "0bitmap$1"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "org.specs2.fp.Show"
+      },
+      "type": "org.specs2.fp.Show$",
+      "fields": [
+        {
+          "name": "0bitmap$1"
+        }
+      ]
+    }
+  ]
+}

--- a/metadata/org.specs2/specs2-fp_3/index.json
+++ b/metadata/org.specs2/specs2-fp_3/index.json
@@ -1,6 +1,19 @@
 [
   {
     "latest" : true,
+    "metadata-version" : "5.0.0",
+    "language" : {
+      "name" : "scala",
+      "version" : "3"
+    },
+    "tested-versions" : [
+      "5.0.0"
+    ],
+    "allowed-packages" : [
+      "org.specs2.fp"
+    ]
+  },
+  {
     "metadata-version" : "4.20.5",
     "source-code-url" : "https://repo.maven.apache.org/maven2/org/specs2/specs2-fp_3/$version$/specs2-fp_3-$version$-sources.jar",
     "repository-url" : "https://github.com/etorreborre/specs2",

--- a/metadata/org.specs2/specs2-fp_3/index.json
+++ b/metadata/org.specs2/specs2-fp_3/index.json
@@ -6,6 +6,11 @@
       "name" : "scala",
       "version" : "3"
     },
+    "source-code-url" : "https://repo.maven.apache.org/maven2/org/specs2/specs2-fp_3/$version$/specs2-fp_3-$version$-sources.jar",
+    "repository-url" : "https://github.com/etorreborre/specs2",
+    "test-code-url" : "https://github.com/etorreborre/specs2/tree/SPECS2-$version$/core/shared/src/test/scala/org/specs2/fp",
+    "documentation-url" : "https://repo.maven.apache.org/maven2/org/specs2/specs2-fp_3/$version$/specs2-fp_3-$version$-javadoc.jar",
+    "description" : "specs2-fp provides the small functional-programming support layer used by specs2, including type classes such as Functor, Monad, Monoid, Semigroup, Traverse, and related syntax/data utilities. It supplies reusable Scala abstractions for composing computations, folding and traversing values, memoization, naming, and tree structures within specs2-based code.",
     "tested-versions" : [
       "5.0.0"
     ],

--- a/stats/org.specs2/specs2-fp_3/5.0.0/execution-metrics.json
+++ b/stats/org.specs2/specs2-fp_3/5.0.0/execution-metrics.json
@@ -1,0 +1,87 @@
+{
+  "fix_java_run_fail:2026-05-10": {
+    "timestamp": "2026-05-10T17:16:26.441839Z",
+    "library": "org.specs2:specs2-fp_3:5.0.0",
+    "starting_commit": "3a704a72bd7338ec99aff1a9f0b8d83b895df380",
+    "ending_commit": "f7dde81f4ae6f49b0f0b1ba3344cea570cece7aa",
+    "previous_library": "org.specs2:specs2-fp_3:4.20.5",
+    "strategy_name": "java_run_iterative_with_coverage_sources_pi_gpt-5.5",
+    "agent": "pi",
+    "model": "oca/gpt-5.5",
+    "status": "success",
+    "stats": {
+      "version": "5.0.0",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 3371,
+          "missed": 3632,
+          "ratio": 0.481365,
+          "total": 7003
+        },
+        "line": {
+          "covered": 339,
+          "missed": 195,
+          "ratio": 0.634831,
+          "total": 534
+        },
+        "method": {
+          "covered": 421,
+          "missed": 467,
+          "ratio": 0.474099,
+          "total": 888
+        }
+      }
+    },
+    "previous_library_stats": {
+      "version": "4.20.5",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 3575,
+          "missed": 3835,
+          "ratio": 0.482456,
+          "total": 7410
+        },
+        "line": {
+          "covered": 381,
+          "missed": 167,
+          "ratio": 0.695255,
+          "total": 548
+        },
+        "method": {
+          "covered": 422,
+          "missed": 517,
+          "ratio": 0.449414,
+          "total": 939
+        }
+      }
+    },
+    "metrics": {
+      "input_tokens_used": 64475,
+      "cached_input_tokens_used": 974848,
+      "output_tokens_used": 11879,
+      "iterations": 1,
+      "cost_usd": 0.8438,
+      "tested_library_loc": 339,
+      "metadata_entries": 7,
+      "previous_library_metadata_entries": 0,
+      "code_coverage_percent": 63.48,
+      "previous_library_coverage_percent": 69.53
+    },
+    "artifacts": {
+      "test_file": "tests/src/org.specs2/specs2-fp_3/5.0.0/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala",
+      "metadata_file": "metadata/org.specs2/specs2-fp_3/5.0.0/reachability-metadata.json"
+    }
+  }
+}

--- a/stats/org.specs2/specs2-fp_3/5.0.0/stats.json
+++ b/stats/org.specs2/specs2-fp_3/5.0.0/stats.json
@@ -1,0 +1,31 @@
+{
+  "versions" : [ {
+    "version" : "5.0.0",
+    "dynamicAccess" : {
+      "breakdown" : { },
+      "coverageRatio" : 1.0,
+      "coveredCalls" : 0,
+      "totalCalls" : 0
+    },
+    "libraryCoverage" : {
+      "instruction" : {
+        "covered" : 3371,
+        "missed" : 3632,
+        "ratio" : 0.481365,
+        "total" : 7003
+      },
+      "line" : {
+        "covered" : 339,
+        "missed" : 195,
+        "ratio" : 0.634831,
+        "total" : 534
+      },
+      "method" : {
+        "covered" : 421,
+        "missed" : 467,
+        "ratio" : 0.474099,
+        "total" : 888
+      }
+    }
+  } ]
+}

--- a/tests/src/org.specs2/specs2-fp_3/5.0.0/.gitignore
+++ b/tests/src/org.specs2/specs2-fp_3/5.0.0/.gitignore
@@ -1,0 +1,4 @@
+gradlew.bat
+gradlew
+gradle/
+build/

--- a/tests/src/org.specs2/specs2-fp_3/5.0.0/build.gradle
+++ b/tests/src/org.specs2/specs2-fp_3/5.0.0/build.gradle
@@ -1,0 +1,36 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+plugins {
+    id "org.graalvm.internal.tck"
+    id "scala"
+}
+
+String libraryVersion = tck.testedLibraryVersion.get()
+
+dependencies {
+    testImplementation "org.specs2:specs2-fp_3:$libraryVersion"
+    testImplementation 'org.assertj:assertj-core:3.22.0'
+    testImplementation 'org.scala-lang:scala3-library_3:3.3.6'
+    testImplementation 'org.scala-lang:scala3-compiler_3:3.3.6'
+}
+
+java {
+    toolchain {
+        languageVersion = JavaLanguageVersion.of(21)
+    }
+}
+
+graalvmNative {
+    agent {
+        defaultMode = "conditional"
+        modes {
+            conditional {
+                userCodeFilterPath = "user-code-filter.json"
+            }
+        }
+    }
+}

--- a/tests/src/org.specs2/specs2-fp_3/5.0.0/gradle.properties
+++ b/tests/src/org.specs2/specs2-fp_3/5.0.0/gradle.properties
@@ -1,0 +1,6 @@
+# Optional explicit tested library coordinates (format: group:artifact:version).
+# If omitted, fallback resolution in org.graalvm.internal.tck.gradle still uses
+# environment variable GVM_TCK_LC and then this property.
+library.coordinates = org.specs2:specs2-fp_3:5.0.0
+library.version = 5.0.0
+metadata.dir = org.specs2/specs2-fp_3/5.0.0/

--- a/tests/src/org.specs2/specs2-fp_3/5.0.0/settings.gradle
+++ b/tests/src/org.specs2/specs2-fp_3/5.0.0/settings.gradle
@@ -1,0 +1,20 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+pluginManagement {
+    def tckPath = Objects.requireNonNullElse(
+            System.getenv("GVM_TCK_TCKDIR"),
+            "../../../../tck-build-logic"
+    )
+    includeBuild(tckPath)
+    repositories { mavenLocal(); gradlePluginPortal() }
+}
+
+plugins {
+    id "org.graalvm.internal.tck-settings" version "1.0.0-SNAPSHOT"
+}
+
+rootProject.name = 'org.specs2.specs2-fp_3_tests'

--- a/tests/src/org.specs2/specs2-fp_3/5.0.0/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala
+++ b/tests/src/org.specs2/specs2-fp_3/5.0.0/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala
@@ -11,9 +11,6 @@ import org.junit.jupiter.api.Assertions.assertFalse
 import org.junit.jupiter.api.Assertions.assertTrue
 import org.junit.jupiter.api.Test
 import org.specs2.fp.Applicative
-import org.specs2.fp.EitherIdOps
-import org.specs2.fp.EitherObjectOps
-import org.specs2.fp.EitherOps
 import org.specs2.fp.Foldable
 import org.specs2.fp.Memo
 import org.specs2.fp.Monad
@@ -30,7 +27,7 @@ import org.specs2.fp.Value
 import org.specs2.fp.syntax.*
 import org.specs2.fp.Id
 
-import scala.collection.immutable.Stream
+import scala.collection.immutable.LazyList
 import scala.concurrent.Await
 import scala.concurrent.ExecutionContext
 import scala.concurrent.Future
@@ -42,13 +39,13 @@ class Specs2_fp_3Test {
 
   @Test
   def monadAndApplicativeInstancesComposeStrictAndOptionalValues(): Unit = {
-    val optionMonad: Monad[Option] = Monad[Option]
+    val optionMonad: Monad[Option] = summon[Monad[Option]]
     val lifted: Option[Int] = optionMonad.point(2)
     val bound: Option[String] = optionMonad.bind(lifted)(i => Some(s"value-${i + 1}"))
     val joined: Option[Int] = optionMonad.join(Some(Some(7)))
-    val tailRecursive: Option[String] = optionMonad.tailrecM[Int, String] { i =>
+    val tailRecursive: Option[String] = optionMonad.tailrecM(1) { i =>
       Some(if i < 4 then Left(i + 1) else Right(s"done-$i"))
-    }(1)
+    }
 
     assertEquals(Some("value-3"), bound)
     assertEquals(Some(7), joined)
@@ -65,13 +62,13 @@ class Specs2_fp_3Test {
     assertEquals(Some((1, "two", true)), tupled)
     assertEquals(Some(6), liftedSum)
     assertEquals(Some(List(2, 4)), filtered)
-    assertEquals(Some(()), optionApplicative.whenM(false)(throw new AssertionError("should not evaluate")))
-    assertEquals(Some(()), optionApplicative.unlessM(true)(throw new AssertionError("should not evaluate")))
+    assertEquals(Some(()), optionApplicative.when(false)(throw new AssertionError("should not evaluate")))
+    assertEquals(Some(()), optionApplicative.unless(true)(throw new AssertionError("should not evaluate")))
   }
 
   @Test
   def monadIterationRepeatsActionsUntilPredicatesChange(): Unit = {
-    val nameMonad: Monad[Name] = Name.name
+    val nameMonad: Monad[Name] = summon[Monad[Name]]
 
     var untilCalls: Int = 0
     val untilValue: Name[Int] = nameMonad.iterateUntil(Name {
@@ -94,8 +91,6 @@ class Specs2_fp_3Test {
 
   @Test
   def syntaxExtensionsDelegateToFunctorApplicativeMonadAndShowTypeClasses(): Unit = {
-    given Applicative[Option] = Applicative.optionApplicative
-
     val mapped: Option[Int] = Option(4).map(_ + 1)
     val replaced: Option[String] = Option(4).as("constant")
     val voided: Option[Unit] = Option(4).void
@@ -156,8 +151,8 @@ class Specs2_fp_3Test {
     assertTrue(foldableList.empty(List.empty[Int]))
     assertEquals("a,b,c", foldableList.intercalate(List("a", "b", "c"), ","))
 
-    val foldedOption: Option[Int] = values.foldLeftM[Option, Int](0)((acc, value) => Some(acc + value))
-    val traversedOption: Option[Unit] = values.traverse_[Option, Int](value => Some(value * 2))
+    val foldedOption: Option[Int] = values.foldLeftM(0)((acc: Int, value: Int) => Option(acc + value))
+    val traversedOption: Option[Unit] = values.traverse_((value: Int) => Option(value * 2))
     assertEquals(Some(10), foldedOption)
     assertEquals(Some(()), traversedOption)
     assertEquals(10, values.sumAll)
@@ -165,16 +160,16 @@ class Specs2_fp_3Test {
 
   @Test
   def traverseInstancesSequenceEffectsAcrossListsOptionsAndEitherValues(): Unit = {
-    val listTraverse: Traverse[List] = Traverse[List]
-    val optionTraverse: Traverse[Option] = Traverse[Option]
-    val eitherTraverse: Traverse[[A] =>> Either[String, A]] = Traverse[[A] =>> Either[String, A]]
+    val listTraverse: Traverse[List] = summon[Traverse[List]]
+    val optionTraverse: Traverse[Option] = summon[Traverse[Option]]
+    val eitherTraverse: Traverse[[A] =>> Either[String, A]] = summon[Traverse[[A] =>> Either[String, A]]]
 
-    val traversedList: Option[List[String]] = listTraverse.traverse(List(1, 2, 3))(i => Option(s"n$i"))
-    val failedList: Option[List[Int]] = listTraverse.traverse(List(1, -1, 3))(i => if i > 0 then Option(i) else None)
+    val traversedList: Option[List[String]] = listTraverse.traverse(List(1, 2, 3))((i: Int) => Option(s"n$i"))
+    val failedList: Option[List[Int]] = listTraverse.traverse(List(1, -1, 3))((i: Int) => if i > 0 then Option(i) else None)
     val sequencedList: Option[List[Int]] = listTraverse.sequence(List[Option[Int]](Some(1), Some(2), Some(3)))
-    val traversedOption: Either[String, Option[Int]] = optionTraverse.traverse(Some(5))(i => Right(i + 1): Either[String, Int])
-    val traversedLeft: Option[Either[String, Int]] = eitherTraverse.traverse(Left("stop"): Either[String, Int])(i => Option(i + 1))
-    val syntaxTraverse: Option[List[Int]] = List(1, 2, 3).traverse[Option, Int](i => Option(i * 10))
+    val traversedOption: Either[String, Option[Int]] = optionTraverse.traverse(Some(5))((i: Int) => Right(i + 1): Either[String, Int])
+    val traversedLeft: Option[Either[String, Int]] = eitherTraverse.traverse(Left("stop"): Either[String, Int])((i: Int) => Option(i + 1))
+    val syntaxTraverse: Option[List[Int]] = List(1, 2, 3).traverse((i: Int) => Option(i * 10))
     val syntaxSequence: Option[List[Int]] = List[Option[Int]](Some(1), Some(2)).sequence
 
     assertEquals(Some(List("n1", "n2", "n3")), traversedList)
@@ -188,43 +183,40 @@ class Specs2_fp_3Test {
 
   @Test
   def eitherSyntaxProvidesCatsLikeOperationsWithoutUsingReflection(): Unit = {
-    val eitherObject: EitherObjectOps = new EitherObjectOps(Either)
-    val rightOps: EitherOps[String, Int] = new EitherOps[String, Int](Right(2))
-    val leftOps: EitherOps[String, Int] = new EitherOps[String, Int](Left("bad"))
-    val numberFormat: Either[Throwable, Int] = eitherObject.catchNonFatal("not-a-number".toInt)
+    val right: Either[String, Int] = Right(2)
+    val left: Either[String, Int] = Left("bad")
+    val eitherApplicative: Applicative[[A] =>> Either[String, A]] = summon[Applicative[[A] =>> Either[String, A]]]
 
-    assertEquals(Right(2), eitherObject.right[String, Int](2))
-    assertEquals(Left("left"), eitherObject.left[String, Int]("left"))
-    assertLeftThrowable(numberFormat, _.isInstanceOf[NumberFormatException])
-    assertEquals(Right(3), eitherObject.fromTry(Success(3)))
-    assertLeftThrowable(eitherObject.fromTry(Failure(new IllegalStateException("boom"))), _.isInstanceOf[IllegalStateException])
-    assertEquals(Right("present"), eitherObject.fromOption(Some("present"), "missing"))
-    assertEquals(Left("missing"), eitherObject.fromOption(None, "missing"))
+    assertEquals(Right(2), Right(2): Either[String, Int])
+    assertEquals(Left("left"), Left("left"): Either[String, Int])
+    assertEquals(Right("present"), Option("present").toRight("missing"))
+    assertEquals(Left("missing"), Option.empty[String].toRight("missing"))
+    assertLeftThrowable(scala.util.Try("not-a-number".toInt).toEither, _.isInstanceOf[NumberFormatException])
+    assertEquals(Right(3), Success(3).toEither)
+    assertLeftThrowable(Failure(new IllegalStateException("boom")).toEither, _.isInstanceOf[IllegalStateException])
 
-    assertEquals(Right(3), rightOps.map(_ + 1))
-    assertEquals(Right("value-2"), rightOps.bimap(_.toUpperCase, i => s"value-$i"))
-    assertEquals(Right(4), rightOps.flatMap(i => Right(i * 2)))
-    assertEquals(Right(5), rightOps.ap(Right((i: Int) => i + 3)))
-    assertEquals(Right(5), rightOps.append(Right(3))(intAdditionSemigroup))
-    assertEquals(Right(2), rightOps.ensure("too-small")(_ >= 2))
-    assertEquals(Some(2), rightOps.toOption)
-    assertEquals(List(2), rightOps.toList)
-    assertEquals(2, rightOps.getOrElse(0))
-    assertEquals(2, rightOps.valueOr(_.length))
-    assertEquals(4, rightOps.foldLeft(2)(_ * _))
-    assertEquals("Right(2)", rightOps.show(Show.showFromToString[String], Show.showFromToString[Int]))
+    assertEquals(Right(3), right.map(_ + 1))
+    assertEquals(Right("value-2"), right.bimap(_.toUpperCase, i => s"value-$i"))
+    assertEquals(Right(4), right.flatMap(i => Right(i * 2)))
+    assertEquals(Right(5), eitherApplicative.ap(right)(Right((i: Int) => i + 3)))
+    assertEquals(Right(5), right.append(Right(3))(using intAdditionSemigroup))
+    assertEquals(Right(2), right.ensure("too-small")(_ >= 2))
+    assertEquals(Some(2), right.toOption)
+    assertEquals(List(2), right.toList)
+    assertEquals(2, right.getOrElse(0))
+    assertEquals(2, right.valueOr(_.length))
+    assertEquals(4, right.foldLeft(2)(_ * _))
+    assertEquals("Right(2)", right.show(using Show.showFromToString[String], Show.showFromToString[Int]))
 
-    assertEquals(Left("BAD"), leftOps.leftMap(_.toUpperCase))
-    assertEquals(Right(7), leftOps.recover { case "bad" => 7 })
-    assertEquals(Right(8), leftOps.recoverWith { case "bad" => Right(8) })
-    assertEquals(Right(9), leftOps.orElse(Right(9)))
-    assertEquals(3, leftOps.valueOr(_.length))
-    assertEquals(Left("bad"), leftOps.append(Right(3))(intAdditionSemigroup))
-    assertEquals(Left("bad"), leftOps.ensure("unused")(_ => false))
-    assertEquals(None, leftOps.toOption)
-    assertEquals(Nil, leftOps.toList)
-    assertEquals(Right(2), new EitherIdOps[Int](2).asRight[String])
-    assertEquals(Left("problem"), new EitherIdOps[String]("problem").asLeft[Int])
+    assertEquals(Left("BAD"), left.leftMap(_.toUpperCase))
+    assertEquals(Right(7), left.recover { case "bad" => 7 })
+    assertEquals(Right(8), left.recoverWith { case "bad" => Right(8) })
+    assertEquals(Right(9), left.orElse(Right(9)))
+    assertEquals(3, left.valueOr(_.length))
+    assertEquals(Left("bad"), left.append(Right(3))(using intAdditionSemigroup))
+    assertEquals(Left("bad"), left.ensure("unused")(_ => false))
+    assertEquals(None, left.toOption)
+    assertEquals(Nil, left.toList)
   }
 
   @Test
@@ -250,7 +242,7 @@ class Specs2_fp_3Test {
     assertEquals("cached", value.value)
     assertEquals("cached", Value.unapply(value).value)
     assertEquals(Some(10), Need.unapply(need))
-    assertEquals(3, Name.name.bind(Name(1))(i => Name(i + 2)).value)
+    assertEquals(3, summon[Monad[Name]].bind(Name(1))(i => Name(i + 2)).value)
 
     var mutableCalls: Int = 0
     val mutableMemoized: Int => String = Memo.mutableHashMapMemo[Int, String].apply { key =>
@@ -283,8 +275,8 @@ class Specs2_fp_3Test {
   def treeSupportsTraversalDrawingMappingBindingAndZipperEditing(): Unit = {
     val tree: Tree[String] = Tree.Node(
       "root",
-      Stream(
-        Tree.Node("left", Stream(Tree.Leaf("left.leaf"))),
+      LazyList(
+        Tree.Node("left", LazyList(Tree.Leaf("left.leaf"))),
         Tree.Leaf("right")
       )
     )
@@ -293,7 +285,7 @@ class Specs2_fp_3Test {
     assertEquals(List("root", "left", "left.leaf", "right"), tree.flatten.toList)
     assertEquals(4, tree.size)
     assertEquals(List(List("root"), List("left", "right"), List("left.leaf")), tree.levels.map(_.toList).toList)
-    assertTrue(tree.drawTree(Show.showFromToString[String]).contains("left.leaf"))
+    assertTrue(tree.drawTree(using Show.showFromToString[String]).contains("left.leaf"))
     assertEquals(List(4, 4, 9, 5), tree.map(_.length).flatten.toList)
     assertEquals(22, tree.foldMap(_.length))
     assertEquals("root/left/left.leaf/right/end", tree.foldRight("end")((label, rest) => s"$label/$rest"))
@@ -301,7 +293,7 @@ class Specs2_fp_3Test {
     val scanned: Tree[Int] = tree.scanr((label, children) => label.length + children.map(_.rootLabel).sum)
     assertEquals(22, scanned.rootLabel)
 
-    val duplicatedLabels: Tree[String] = tree.flatMap(label => Tree.Node(label.toUpperCase, Stream(Tree.Leaf(label.toLowerCase))))
+    val duplicatedLabels: Tree[String] = tree.flatMap(label => Tree.Node(label.toUpperCase, LazyList(Tree.Leaf(label.toLowerCase))))
     assertEquals(List("ROOT", "root", "LEFT", "left", "LEFT.LEAF", "left.leaf", "RIGHT", "right"), duplicatedLabels.flatten.toList)
 
     val loc: TreeLoc[String] = tree.loc
@@ -315,12 +307,12 @@ class Specs2_fp_3Test {
     assertTrue(loc.isRoot)
     assertTrue(leftChild.isChild)
     assertTrue(leftChild.hasChildren)
-    assertEquals(Stream("left", "root").toList, leftChild.path.toList)
+    assertEquals(LazyList("left", "root").toList, leftChild.path.toList)
     assertEquals("right", rightChild.getLabel)
     assertEquals(List("root", "left", "left.leaf", "middle", "right-updated"), editedRoot.toTree.flatten.toList)
     assertEquals(Some("left"), loc.findChild(_.rootLabel == "left").map(_.getLabel))
     assertEquals(Some("left.leaf"), loc.find(_.getLabel == "left.leaf").map(_.getLabel))
-    assertEquals(Some("root"), TreeLoc.fromForest(Stream(tree)).map(_.getLabel))
+    assertEquals(List("root"), loc.toForest.map(_.rootLabel).toList)
     assertEquals(List(4, 4, 9, 5), loc.map(_.length).toTree.flatten.toList)
   }
 
@@ -328,15 +320,15 @@ class Specs2_fp_3Test {
   def treeCobindAndUnzipCreateContextAwareAndSplitTrees(): Unit = {
     val tree: Tree[String] = Tree.Node(
       "root",
-      Stream(
-        Tree.Node("left", Stream(Tree.Leaf("left.leaf"))),
+      LazyList(
+        Tree.Node("left", LazyList(Tree.Leaf("left.leaf"))),
         Tree.Leaf("right")
       )
     )
 
     val contextSizes: Tree[String] = tree.cobind(subtree => s"${subtree.rootLabel}:${subtree.size}")
     val labelled: Tree[(String, Int)] = tree.map(label => (label, label.length))
-    val unzipped: (Tree[String], Tree[Int]) = labelled.unzip { case (label, length) => (label, length) }
+    val unzipped: (Tree[String], Tree[Int]) = labelled.unzip(using (pair: (String, Int)) => pair)
 
     assertEquals(List("root:4", "left:2", "left.leaf:1", "right:1"), contextSizes.flatten.toList)
     assertEquals(List("root", "left", "left.leaf", "right"), unzipped._1.flatten.toList)
@@ -346,13 +338,13 @@ class Specs2_fp_3Test {
   @Test
   def unfoldBuildsTreesAndForestsFromSeeds(): Unit = {
     val unfolded: Tree[Int] = Tree.unfoldTree(1) { seed =>
-      val children: Stream[Int] = if seed < 3 then Stream(seed + 1, seed + 2) else Stream.empty
+      val children: LazyList[Int] = if seed < 3 then LazyList(seed + 1, seed + 2) else LazyList.empty
       (seed, () => children)
     }
-    val unfoldedForest: Stream[Tree[Int]] = Tree.unfoldForest(Stream(1, 2)) { seed =>
-      (seed * 10, () => Stream.empty[Int])
+    val unfoldedForest: LazyList[Tree[Int]] = Tree.unfoldForest(LazyList(1, 2)) { seed =>
+      (seed * 10, () => LazyList.empty[Int])
     }
-    val counted: Stream[Int] = TreeLoc.unfold(0)(i => if i < 3 then Some((i + 1, i + 1)) else None)
+    val counted: LazyList[Int] = TreeLoc.unfold(0)(i => if i < 3 then Some((i + 1, i + 1)) else None)
 
     assertEquals(List(1, 2, 3, 4, 3), unfolded.flatten.toList)
     assertEquals(List(10, 20), unfoldedForest.map(_.rootLabel).toList)
@@ -363,15 +355,15 @@ class Specs2_fp_3Test {
   def futureInstancesAndNaturalTransformationsProduceExpectedValuesWithBoundedAwait(): Unit = {
     given ExecutionContext = ExecutionContext.fromExecutor((command: Runnable) => command.run())
 
-    val futureMonad: Monad[Future] = Monad.futureMonad
-    val futureApplicative: Applicative[Future] = Applicative.futureApplicative
+    val futureMonad: Monad[Future] = summon[Monad[Future]]
+    val futureApplicative: Applicative[Future] = summon[Applicative[Future]]
     val futureValue: Future[Int] = futureMonad.bind(Future.successful(4))(i => Future.successful(i * 3))
     val futureTuple: Future[(String, Int)] = futureApplicative.tuple2(Future.successful("age"), Future.successful(42))
 
     assertEquals(12, Await.result(futureValue, 5.seconds))
     assertEquals(("age", 42), Await.result(futureTuple, 5.seconds))
 
-    val natural: NaturalTransformation[Id, Option] = NaturalTransformation.naturalId[Option]
+    val natural: NaturalTransformation[Id, Option] = summon[NaturalTransformation[Id, Option]]
     assertEquals(Some("lifted"), natural.apply("lifted"))
   }
 

--- a/tests/src/org.specs2/specs2-fp_3/5.0.0/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala
+++ b/tests/src/org.specs2/specs2-fp_3/5.0.0/src/test/scala/org_specs2/specs2_fp_3/Specs2_fp_3Test.scala
@@ -1,0 +1,390 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package org_specs2.specs2_fp_3
+
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertFalse
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.Test
+import org.specs2.fp.Applicative
+import org.specs2.fp.EitherIdOps
+import org.specs2.fp.EitherObjectOps
+import org.specs2.fp.EitherOps
+import org.specs2.fp.Foldable
+import org.specs2.fp.Memo
+import org.specs2.fp.Monad
+import org.specs2.fp.Monoid
+import org.specs2.fp.Name
+import org.specs2.fp.NaturalTransformation
+import org.specs2.fp.Need
+import org.specs2.fp.Semigroup
+import org.specs2.fp.Show
+import org.specs2.fp.Traverse
+import org.specs2.fp.Tree
+import org.specs2.fp.TreeLoc
+import org.specs2.fp.Value
+import org.specs2.fp.syntax.*
+import org.specs2.fp.Id
+
+import scala.collection.immutable.Stream
+import scala.concurrent.Await
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.duration.DurationInt
+import scala.util.Failure
+import scala.util.Success
+
+class Specs2_fp_3Test {
+
+  @Test
+  def monadAndApplicativeInstancesComposeStrictAndOptionalValues(): Unit = {
+    val optionMonad: Monad[Option] = Monad[Option]
+    val lifted: Option[Int] = optionMonad.point(2)
+    val bound: Option[String] = optionMonad.bind(lifted)(i => Some(s"value-${i + 1}"))
+    val joined: Option[Int] = optionMonad.join(Some(Some(7)))
+    val tailRecursive: Option[String] = optionMonad.tailrecM[Int, String] { i =>
+      Some(if i < 4 then Left(i + 1) else Right(s"done-$i"))
+    }(1)
+
+    assertEquals(Some("value-3"), bound)
+    assertEquals(Some(7), joined)
+    assertEquals(Some("done-4"), tailRecursive)
+    assertEquals(None, optionMonad.bind(None: Option[Int])(_ => Some("unreachable")))
+
+    val optionApplicative: Applicative[Option] = Applicative[Option]
+    val applied: Option[Int] = optionApplicative.ap(Some(3))(Some((i: Int) => i + 4))
+    val tupled: Option[(Int, String, Boolean)] = optionApplicative.tuple3(Some(1), Some("two"), Some(true))
+    val liftedSum: Option[Int] = optionApplicative.lift3((a: Int, b: Int, c: Int) => a + b + c)(Some(1), Some(2), Some(3))
+    val filtered: Option[List[Int]] = optionApplicative.filterM(List(1, 2, 3, 4))(i => Some(i % 2 == 0))
+
+    assertEquals(Some(7), applied)
+    assertEquals(Some((1, "two", true)), tupled)
+    assertEquals(Some(6), liftedSum)
+    assertEquals(Some(List(2, 4)), filtered)
+    assertEquals(Some(()), optionApplicative.whenM(false)(throw new AssertionError("should not evaluate")))
+    assertEquals(Some(()), optionApplicative.unlessM(true)(throw new AssertionError("should not evaluate")))
+  }
+
+  @Test
+  def monadIterationRepeatsActionsUntilPredicatesChange(): Unit = {
+    val nameMonad: Monad[Name] = Name.name
+
+    var untilCalls: Int = 0
+    val untilValue: Name[Int] = nameMonad.iterateUntil(Name {
+      untilCalls += 1
+      untilCalls
+    })(_ >= 3)
+
+    assertEquals(3, untilValue.value)
+    assertEquals(3, untilCalls)
+
+    var whileCalls: Int = 0
+    val whileValue: Name[Int] = nameMonad.iterateWhile(Name {
+      whileCalls += 1
+      whileCalls
+    })(_ < 4)
+
+    assertEquals(4, whileValue.value)
+    assertEquals(4, whileCalls)
+  }
+
+  @Test
+  def syntaxExtensionsDelegateToFunctorApplicativeMonadAndShowTypeClasses(): Unit = {
+    given Applicative[Option] = Applicative.optionApplicative
+
+    val mapped: Option[Int] = Option(4).map(_ + 1)
+    val replaced: Option[String] = Option(4).as("constant")
+    val voided: Option[Unit] = Option(4).void
+    val combined: Option[String] = Option(2).|@|(Option(3))((a, b) => s"$a+$b=${a + b}")
+    val second: Option[String] = Option(1).*>(Option("kept"))
+    val monadic: Option[Int] = Option(3).>>=(i => Some(i * 2))
+
+    assertEquals(Some(5), mapped)
+    assertEquals(Some("constant"), replaced)
+    assertEquals(Some(()), voided)
+    assertEquals(Some("2+3=5"), combined)
+    assertEquals(Some("kept"), second)
+    assertEquals(Some(6), monadic)
+
+    given Show[String] = Show.show(value => s"<$value>")
+    assertEquals("<visible>", "visible".show)
+    assertEquals("42", Show[Int].show(42))
+  }
+
+  @Test
+  def semigroupsAndMonoidsAppendMultiplyAndMergeCollections(): Unit = {
+    given Semigroup[String] = Monoid.stringMonoid
+
+    val commaSeparated: Semigroup[String] = Semigroup.instance((left, right) => s"$left,$right")
+    val stringMonoid: Monoid[String] = Monoid[String]
+    val intMonoid: Monoid[Int] = Monoid[Int]
+    val listMonoid: Monoid[List[Int]] = Monoid[List[Int]]
+    val mapMonoid: Monoid[Map[String, Int]] = Monoid.mapMonoid[String, Int]
+
+    assertEquals("a,b", commaSeparated.append("a", "b"))
+    assertEquals("x,x,x", commaSeparated.multiply1("x", 2))
+    assertEquals("ababab", stringMonoid.multiply("ab", 3))
+    assertEquals(0, intMonoid.zero)
+    assertEquals(10, intMonoid.append(4, 6))
+    assertEquals(List(1, 2, 3), listMonoid.append(List(1), List(2, 3)))
+    assertEquals(Map("a" -> 3, "b" -> 5), mapMonoid.append(Map("a" -> 1), Map("a" -> 2, "b" -> 5)))
+    assertEquals("leftright", "left".|+|("right"))
+  }
+
+  @Test
+  def foldableInstancesFoldSearchConvertAndTraverseCollections(): Unit = {
+    val foldableList: Foldable[List] = Foldable[List]
+    val values: List[Int] = List(1, 2, 3, 4)
+
+    assertEquals(10, foldableList.foldMap(values)(identity))
+    assertEquals("start-1-2-3-4", foldableList.foldLeft(values, "start")((acc, i) => s"$acc-$i"))
+    assertEquals(Some(2), foldableList.findLeft(values)(_ % 2 == 0))
+    assertEquals(Some(3), foldableList.findRight(values)(_ % 2 == 1))
+    assertEquals(4, foldableList.count(values))
+    assertEquals(Some(3), foldableList.index(values, 2))
+    assertEquals(99, foldableList.indexOr(values, 99, 40))
+    assertEquals(Vector(1, 2, 3, 4), foldableList.toVector(values))
+    assertEquals(Set(1, 2, 3, 4), foldableList.toSet(values))
+    assertEquals(List(1, 2, 3, 4), foldableList.toStream(values).toList)
+    assertTrue(foldableList.all(values)(_ > 0))
+    assertTrue(foldableList.any(values)(_ == 3))
+    assertFalse(foldableList.empty(values))
+    assertTrue(foldableList.empty(List.empty[Int]))
+    assertEquals("a,b,c", foldableList.intercalate(List("a", "b", "c"), ","))
+
+    val foldedOption: Option[Int] = values.foldLeftM[Option, Int](0)((acc, value) => Some(acc + value))
+    val traversedOption: Option[Unit] = values.traverse_[Option, Int](value => Some(value * 2))
+    assertEquals(Some(10), foldedOption)
+    assertEquals(Some(()), traversedOption)
+    assertEquals(10, values.sumAll)
+  }
+
+  @Test
+  def traverseInstancesSequenceEffectsAcrossListsOptionsAndEitherValues(): Unit = {
+    val listTraverse: Traverse[List] = Traverse[List]
+    val optionTraverse: Traverse[Option] = Traverse[Option]
+    val eitherTraverse: Traverse[[A] =>> Either[String, A]] = Traverse[[A] =>> Either[String, A]]
+
+    val traversedList: Option[List[String]] = listTraverse.traverse(List(1, 2, 3))(i => Option(s"n$i"))
+    val failedList: Option[List[Int]] = listTraverse.traverse(List(1, -1, 3))(i => if i > 0 then Option(i) else None)
+    val sequencedList: Option[List[Int]] = listTraverse.sequence(List[Option[Int]](Some(1), Some(2), Some(3)))
+    val traversedOption: Either[String, Option[Int]] = optionTraverse.traverse(Some(5))(i => Right(i + 1): Either[String, Int])
+    val traversedLeft: Option[Either[String, Int]] = eitherTraverse.traverse(Left("stop"): Either[String, Int])(i => Option(i + 1))
+    val syntaxTraverse: Option[List[Int]] = List(1, 2, 3).traverse[Option, Int](i => Option(i * 10))
+    val syntaxSequence: Option[List[Int]] = List[Option[Int]](Some(1), Some(2)).sequence
+
+    assertEquals(Some(List("n1", "n2", "n3")), traversedList)
+    assertEquals(None, failedList)
+    assertEquals(Some(List(1, 2, 3)), sequencedList)
+    assertEquals(Right(Some(6)), traversedOption)
+    assertEquals(Some(Left("stop")), traversedLeft)
+    assertEquals(Some(List(10, 20, 30)), syntaxTraverse)
+    assertEquals(Some(List(1, 2)), syntaxSequence)
+  }
+
+  @Test
+  def eitherSyntaxProvidesCatsLikeOperationsWithoutUsingReflection(): Unit = {
+    val eitherObject: EitherObjectOps = new EitherObjectOps(Either)
+    val rightOps: EitherOps[String, Int] = new EitherOps[String, Int](Right(2))
+    val leftOps: EitherOps[String, Int] = new EitherOps[String, Int](Left("bad"))
+    val numberFormat: Either[Throwable, Int] = eitherObject.catchNonFatal("not-a-number".toInt)
+
+    assertEquals(Right(2), eitherObject.right[String, Int](2))
+    assertEquals(Left("left"), eitherObject.left[String, Int]("left"))
+    assertLeftThrowable(numberFormat, _.isInstanceOf[NumberFormatException])
+    assertEquals(Right(3), eitherObject.fromTry(Success(3)))
+    assertLeftThrowable(eitherObject.fromTry(Failure(new IllegalStateException("boom"))), _.isInstanceOf[IllegalStateException])
+    assertEquals(Right("present"), eitherObject.fromOption(Some("present"), "missing"))
+    assertEquals(Left("missing"), eitherObject.fromOption(None, "missing"))
+
+    assertEquals(Right(3), rightOps.map(_ + 1))
+    assertEquals(Right("value-2"), rightOps.bimap(_.toUpperCase, i => s"value-$i"))
+    assertEquals(Right(4), rightOps.flatMap(i => Right(i * 2)))
+    assertEquals(Right(5), rightOps.ap(Right((i: Int) => i + 3)))
+    assertEquals(Right(5), rightOps.append(Right(3))(intAdditionSemigroup))
+    assertEquals(Right(2), rightOps.ensure("too-small")(_ >= 2))
+    assertEquals(Some(2), rightOps.toOption)
+    assertEquals(List(2), rightOps.toList)
+    assertEquals(2, rightOps.getOrElse(0))
+    assertEquals(2, rightOps.valueOr(_.length))
+    assertEquals(4, rightOps.foldLeft(2)(_ * _))
+    assertEquals("Right(2)", rightOps.show(Show.showFromToString[String], Show.showFromToString[Int]))
+
+    assertEquals(Left("BAD"), leftOps.leftMap(_.toUpperCase))
+    assertEquals(Right(7), leftOps.recover { case "bad" => 7 })
+    assertEquals(Right(8), leftOps.recoverWith { case "bad" => Right(8) })
+    assertEquals(Right(9), leftOps.orElse(Right(9)))
+    assertEquals(3, leftOps.valueOr(_.length))
+    assertEquals(Left("bad"), leftOps.append(Right(3))(intAdditionSemigroup))
+    assertEquals(Left("bad"), leftOps.ensure("unused")(_ => false))
+    assertEquals(None, leftOps.toOption)
+    assertEquals(Nil, leftOps.toList)
+    assertEquals(Right(2), new EitherIdOps[Int](2).asRight[String])
+    assertEquals(Left("problem"), new EitherIdOps[String]("problem").asLeft[Int])
+  }
+
+  @Test
+  def nameNeedValueAndMemoControlEvaluationAndCaching(): Unit = {
+    var nameEvaluations: Int = 0
+    val name: Name[Int] = Name {
+      nameEvaluations += 1
+      nameEvaluations
+    }
+    assertEquals(1, name.value)
+    assertEquals(2, name.value)
+
+    var needEvaluations: Int = 0
+    val need: Need[Int] = Need {
+      needEvaluations += 1
+      needEvaluations * 10
+    }
+    assertEquals(10, need.value)
+    assertEquals(10, need.value)
+    assertEquals(1, needEvaluations)
+
+    val value: Value[String] = Value("cached")
+    assertEquals("cached", value.value)
+    assertEquals("cached", Value.unapply(value).value)
+    assertEquals(Some(10), Need.unapply(need))
+    assertEquals(3, Name.name.bind(Name(1))(i => Name(i + 2)).value)
+
+    var mutableCalls: Int = 0
+    val mutableMemoized: Int => String = Memo.mutableHashMapMemo[Int, String].apply { key =>
+      mutableCalls += 1
+      s"value-$key-$mutableCalls"
+    }
+    assertEquals("value-1-1", mutableMemoized(1))
+    assertEquals("value-1-1", mutableMemoized(1))
+    assertEquals(1, mutableCalls)
+
+    var immutableCalls: Int = 0
+    val immutableMemoized: Int => Int = Memo.immutableTreeMapMemo[Int, Int].apply { key =>
+      immutableCalls += 1
+      key * 10 + immutableCalls
+    }
+    assertEquals(21, immutableMemoized(2))
+    assertEquals(21, immutableMemoized(2))
+    assertEquals(1, immutableCalls)
+
+    var nilCalls: Int = 0
+    val notMemoized: Int => Int = Memo.nilMemo[Int, Int].apply { key =>
+      nilCalls += 1
+      key + nilCalls
+    }
+    assertEquals(6, notMemoized(5))
+    assertEquals(7, notMemoized(5))
+  }
+
+  @Test
+  def treeSupportsTraversalDrawingMappingBindingAndZipperEditing(): Unit = {
+    val tree: Tree[String] = Tree.Node(
+      "root",
+      Stream(
+        Tree.Node("left", Stream(Tree.Leaf("left.leaf"))),
+        Tree.Leaf("right")
+      )
+    )
+
+    assertEquals("root", tree.rootLabel)
+    assertEquals(List("root", "left", "left.leaf", "right"), tree.flatten.toList)
+    assertEquals(4, tree.size)
+    assertEquals(List(List("root"), List("left", "right"), List("left.leaf")), tree.levels.map(_.toList).toList)
+    assertTrue(tree.drawTree(Show.showFromToString[String]).contains("left.leaf"))
+    assertEquals(List(4, 4, 9, 5), tree.map(_.length).flatten.toList)
+    assertEquals(22, tree.foldMap(_.length))
+    assertEquals("root/left/left.leaf/right/end", tree.foldRight("end")((label, rest) => s"$label/$rest"))
+
+    val scanned: Tree[Int] = tree.scanr((label, children) => label.length + children.map(_.rootLabel).sum)
+    assertEquals(22, scanned.rootLabel)
+
+    val duplicatedLabels: Tree[String] = tree.flatMap(label => Tree.Node(label.toUpperCase, Stream(Tree.Leaf(label.toLowerCase))))
+    assertEquals(List("ROOT", "root", "LEFT", "left", "LEFT.LEAF", "left.leaf", "RIGHT", "right"), duplicatedLabels.flatten.toList)
+
+    val loc: TreeLoc[String] = tree.loc
+    val leftChild: TreeLoc[String] = loc.firstChild.getOrElse(failLocation("missing left child"))
+    val rightChild: TreeLoc[String] = leftChild.right.getOrElse(failLocation("missing right child"))
+    val editedRoot: TreeLoc[String] = rightChild
+      .setLabel("right-updated")
+      .insertLeft(Tree.Leaf("middle"))
+      .root
+
+    assertTrue(loc.isRoot)
+    assertTrue(leftChild.isChild)
+    assertTrue(leftChild.hasChildren)
+    assertEquals(Stream("left", "root").toList, leftChild.path.toList)
+    assertEquals("right", rightChild.getLabel)
+    assertEquals(List("root", "left", "left.leaf", "middle", "right-updated"), editedRoot.toTree.flatten.toList)
+    assertEquals(Some("left"), loc.findChild(_.rootLabel == "left").map(_.getLabel))
+    assertEquals(Some("left.leaf"), loc.find(_.getLabel == "left.leaf").map(_.getLabel))
+    assertEquals(Some("root"), TreeLoc.fromForest(Stream(tree)).map(_.getLabel))
+    assertEquals(List(4, 4, 9, 5), loc.map(_.length).toTree.flatten.toList)
+  }
+
+  @Test
+  def treeCobindAndUnzipCreateContextAwareAndSplitTrees(): Unit = {
+    val tree: Tree[String] = Tree.Node(
+      "root",
+      Stream(
+        Tree.Node("left", Stream(Tree.Leaf("left.leaf"))),
+        Tree.Leaf("right")
+      )
+    )
+
+    val contextSizes: Tree[String] = tree.cobind(subtree => s"${subtree.rootLabel}:${subtree.size}")
+    val labelled: Tree[(String, Int)] = tree.map(label => (label, label.length))
+    val unzipped: (Tree[String], Tree[Int]) = labelled.unzip { case (label, length) => (label, length) }
+
+    assertEquals(List("root:4", "left:2", "left.leaf:1", "right:1"), contextSizes.flatten.toList)
+    assertEquals(List("root", "left", "left.leaf", "right"), unzipped._1.flatten.toList)
+    assertEquals(List(4, 4, 9, 5), unzipped._2.flatten.toList)
+  }
+
+  @Test
+  def unfoldBuildsTreesAndForestsFromSeeds(): Unit = {
+    val unfolded: Tree[Int] = Tree.unfoldTree(1) { seed =>
+      val children: Stream[Int] = if seed < 3 then Stream(seed + 1, seed + 2) else Stream.empty
+      (seed, () => children)
+    }
+    val unfoldedForest: Stream[Tree[Int]] = Tree.unfoldForest(Stream(1, 2)) { seed =>
+      (seed * 10, () => Stream.empty[Int])
+    }
+    val counted: Stream[Int] = TreeLoc.unfold(0)(i => if i < 3 then Some((i + 1, i + 1)) else None)
+
+    assertEquals(List(1, 2, 3, 4, 3), unfolded.flatten.toList)
+    assertEquals(List(10, 20), unfoldedForest.map(_.rootLabel).toList)
+    assertEquals(List(1, 2, 3), counted.toList)
+  }
+
+  @Test
+  def futureInstancesAndNaturalTransformationsProduceExpectedValuesWithBoundedAwait(): Unit = {
+    given ExecutionContext = ExecutionContext.fromExecutor((command: Runnable) => command.run())
+
+    val futureMonad: Monad[Future] = Monad.futureMonad
+    val futureApplicative: Applicative[Future] = Applicative.futureApplicative
+    val futureValue: Future[Int] = futureMonad.bind(Future.successful(4))(i => Future.successful(i * 3))
+    val futureTuple: Future[(String, Int)] = futureApplicative.tuple2(Future.successful("age"), Future.successful(42))
+
+    assertEquals(12, Await.result(futureValue, 5.seconds))
+    assertEquals(("age", 42), Await.result(futureTuple, 5.seconds))
+
+    val natural: NaturalTransformation[Id, Option] = NaturalTransformation.naturalId[Option]
+    assertEquals(Some("lifted"), natural.apply("lifted"))
+  }
+
+  private val intAdditionSemigroup: Semigroup[Int] = new Semigroup[Int] {
+    override def append(left: Int, right: => Int): Int = left + right
+  }
+
+  private def assertLeftThrowable(either: Either[Throwable, ?], expected: Throwable => Boolean): Unit =
+    either match {
+      case Left(error) => assertTrue(expected(error), s"unexpected error: $error")
+      case Right(value) => throw new AssertionError(s"expected a Left but got Right($value)")
+    }
+
+  private def failLocation[A](message: String): A =
+    throw new AssertionError(message)
+}

--- a/tests/src/org.specs2/specs2-fp_3/5.0.0/user-code-filter.json
+++ b/tests/src/org.specs2/specs2-fp_3/5.0.0/user-code-filter.json
@@ -1,0 +1,13 @@
+{
+  "rules" : [
+    {
+      "excludeClasses" : "**"
+    },
+    {
+      "includeClasses" : "org.specs2.fp.**"
+    },
+    {
+      "includeClasses" : "org_specs2.specs2_fp_3.**"
+    }
+  ]
+}

